pkg/generate/code/resource_reference.go

```diff
@@ -179,6 +185,187 @@ func ReferenceFieldsPresent(
+//				"Integration",
+//				namespace, *arr.Name,
+//				"Status.IntegrationID")
+//		}
+//		refVal = string(*obj.Status.IntegrationID)
+//	}
+//
+// ko.Spec.Target = &refVal
+// ```
+func ResolveReferencesForField(r *model.CRD, field *model.Field, sourceVarName string, indentLevel int) string {
+	fp := fieldpath.FromString(field.Path)
+
+	outPrefix := ""
+	outSuffix := ""
+
+	fieldAccessPrefix := fmt.Sprintf("%s%s", sourceVarName, r.Config().PrefixConfig.SpecField)
+	targetVarName := fmt.Sprintf("%s%s.%s", sourceVarName, r.Config().PrefixConfig.SpecField, field.Path)
+
+	for idx := 0; idx < fp.Size(); idx++ {
+		curFP := fp.CopyAt(idx).String()
+		cur, ok := r.Fields[curFP]
+		if !ok {
+			panic(fmt.Sprintf("unable to find field with path %s", curFP))
+		}
+
+		ref := cur.ShapeRef
+
+		indent := strings.Repeat("\t", indentLevel+idx)
+
+		if ref.Shape.Type == "structure" {
+			fieldAccessPrefix = fmt.Sprintf("%s.%s", fieldAccessPrefix, fp.At(idx))
+
+			outPrefix += fmt.Sprintf("%sif %s != nil {\n", indent, fieldAccessPrefix)
+			outSuffix = fmt.Sprintf("%s}\n%s", indent, outSuffix)
+		} else if ref.Shape.Type == "list" {
+			if (fp.Size() - idx) > 1 {
+				// TODO(nithomso): add support for structs nested within lists
+				// The logic for structs nested within lists needs to not only
+				// be added here, but also in a custom patching solution since
+				// it isn't supported by `StrategicMergePatch`
+				// see https://github.com/aws-controllers-k8s/community/issues/1291
+				panic("references within lists inside lists aren't supported")
+			}
+			fieldAccessPrefix = fmt.Sprintf("%s.%s", fieldAccessPrefix, fp.At(idx))
+
+			iterVarName := fmt.Sprintf("iter%d", idx)
+			refsTarget := "refVals"
+
+			// base case for references in a list
+			outPrefix += fmt.Sprintf("%s%s := []*string{}\n", indent, refsTarget)
+			outPrefix += fmt.Sprintf("%sfor _, %s := range %s {\n", indent, iterVarName, fieldAccessPrefix)
+
+			fieldAccessPrefix = iterVarName
+			outPrefix += resolveSingleReference(field, fieldAccessPrefix, refsTarget, true, indentLevel+idx+1)
+			outSuffix = fmt.Sprintf("%s%s = %s\n%s", indent, targetVarName, refsTarget, outSuffix)
+			outSuffix = fmt.Sprintf("%s}\n%s", indent, outSuffix)
+		} else if ref.Shape.Type == "map" {
+			panic("references cannot be within a map")
+		} else {
+			// base case for single references
+			refTarget := "refVal"
+			fieldAccessPrefix = fmt.Sprintf("%s.%s", fieldAccessPrefix, cur.GetReferenceFieldName().Camel)
+
+			outPrefix += fmt.Sprintf("%s%s := \"\"\n", indent, refTarget)
+			outPrefix += resolveSingleReference(field, fieldAccessPrefix, refTarget, false, indentLevel+idx)
+			outPrefix += fmt.Sprintf("%s%s = &%s\n", indent, targetVarName, refTarget)
+		}
+	}
+
+	return outPrefix + outSuffix
+}
+
+func resolveSingleReference(field *model.Field, sourceVarName string, targetVarName string, shouldAppendTarget bool, indentLevel int) string {
+	out := ""
+	indent := strings.Repeat("\t", indentLevel)
```

Comment:
Can you just use https://pkg.go.dev/go/format#Source to deal with all of the formatting of the end result instead of bothering will all of these indentations?

Reply:
++ Absolutely would love to do this programmatically. If we can do this as a full refactor for all methods in the code-generator.